### PR TITLE
CMR-7356 + CMR-7355 + CMR-7251

### PR DESCRIPTION
**Pull request by mitchstartzel**
CMR-7356: Reject subscription ingests with invalid query parameters
CMR-7355: Subscription documentation revisions/cleanup
CMR-7251: Improved subscription processing job logging

### CODE REVIEW

**Review by jaybarra — CHANGES_REQUESTED** [review]

File: common-app-lib/test/cmr/common_app/test/services/ingest/subscription.clj [hunk 1]
```diff
@@ -20,3 +20,8 @@
          expected "d41d8cd98f00b204e9800998ecf8427e"
          actual (sub-common/normalize-parameters query)]
      (is (= expected actual)))))
+
+(deftest create-query-params
+  (is (= {"polygon" "-78,-18,-77,-22,-73,-16,-74,-13,-78,-18"
+          "concept-id" "G123-PROV1"}
+         (#'sub-common/create-query-params "polygon=-78,-18,-77,-22,-73,-16,-74,-13,-78,-18&concept-id=G123-PROV1"))))
```
Comment (ygliuvt) [inline common-app-lib/test/cmr/common_app/test/services/ingest/subscription.clj:27]: `create-query-params` is public now, you can remove the `#'`

File: ingest-app/src/cmr/ingest/api/subscriptions.clj [hunk 2]
```diff
@@ -224,12 +248,13 @@
                             content-type
                             headers)
           native-id (get-unique-native-id request-context tmp-subscription)
-          new-subscription (assoc  tmp-subscription :native-id native-id)
-          newer-subscription (add-fields-if-missing request-context new-subscription)
-          subscriber-id (get-subscriber-id newer-subscription)]
+          sub-with-native-id (assoc  tmp-subscription :native-id native-id)
+          final-sub (add-fields-if-missing request-context sub-with-native-id)
+          subscriber-id (get-subscriber-id final-sub)]
+      (validate-query request-context final-sub)
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/api/subscriptions.clj:254]: Move the `validate-query` call below `check-ingest-permission` call.

File: ingest-app/src/cmr/ingest/api/subscriptions.clj [hunk 3]
```diff
@@ -224,12 +248,13 @@
                             content-type
                             headers)
           native-id (get-unique-native-id request-context tmp-subscription)
-          new-subscription (assoc  tmp-subscription :native-id native-id)
-          newer-subscription (add-fields-if-missing request-context new-subscription)
-          subscriber-id (get-subscriber-id newer-subscription)]
+          sub-with-native-id (assoc  tmp-subscription :native-id native-id)
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/api/subscriptions.clj:251]: remove the extra space before `tmp-subscription`.

File: ingest-app/src/cmr/ingest/api/subscriptions.clj [hunk 4]
```diff
@@ -249,11 +274,12 @@
                             body
                             content-type
                             headers)
-          new-subscription (add-fields-if-missing request-context tmp-subscription)
-          subscriber-id (get-subscriber-id new-subscription)]
+          final-sub (add-fields-if-missing request-context tmp-subscription)
+          subscriber-id (get-subscriber-id final-sub)]
+      (validate-query request-context final-sub)
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/api/subscriptions.clj:279]: same as above. move this call down.

File: ingest-app/src/cmr/ingest/api/subscriptions.clj [hunk 5]
```diff
@@ -60,6 +62,28 @@
...
            concept-id)))))
 
+(defn- search-gran-refs-with-sub-params
+  "Performs a granule search using the provided query parameters. If the query is no good,
+   we throw a service error."
+  [context params]
+  (try
+    (search/find-granule-references context params)
+    (catch Exception e
+      (errors/throw-service-error :bad-request
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/api/subscriptions.clj:72]: The error should contain the specific error message returned from the search-app.

File: ingest-app/src/cmr/ingest/api/subscriptions.clj [hunk 6]
```diff
@@ -60,6 +62,28 @@
...
+       (str "Error ingesting subscription: invalid query parameters [" (dissoc params :token) "]")))))
+
+(defn- validate-query
+  "Performs a granule search using subscription query parameters for purposes of validation"
+  [request-context subscription]
+  (let [metadata (-> (:metadata subscription) (json/decode true))
+        collection-id (:CollectionConceptId metadata)
+        query-string (:Query metadata)
+        query-params (sub-common/create-query-params query-string)
+        search-params (merge {:collection-concept-id collection-id
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/api/subscriptions.clj:82]: add a `:page-size 0` to limit the request size.

**Review by ygliuvt — CHANGES_REQUESTED** [review]
Need to address the review comments.

File: ingest-app/src/cmr/ingest/api/subscriptions.clj (hunk 6 above)
Comment (mitchstartzel) [reply ingest-app/src/cmr/ingest/api/subscriptions.clj:82]: Now this is smart.

File: ingest-app/src/cmr/ingest/api/subscriptions.clj (hunk 2 above)
Comment (mitchstartzel) [reply ingest-app/src/cmr/ingest/api/subscriptions.clj]: I like this change - my original thought process was that if the query is bad, it's not worth wasting time checking anything else, so we look for that first. But the same is true if the user has no permissions, which is cheaper to check, so I agree it should go first.

File: common-app-lib/src/cmr/common_app/services/ingest/subscription_common.clj [hunk 7]
```diff
@@ -18,3 +18,14 @@
         (as-> $ (string/join "&" $))
         string/trim
         digest/md5)))
+
+(defn create-query-params
```
Comment (eereiter) [inline common-app-lib/src/cmr/common_app/services/ingest/subscription_common.clj:22]: why was this moved into common?  Is another micro service other than ingest going to use it?  right now I only see ingest using it.   If so then great, if not it probably should remain in ingest.

File: ingest-app/src/cmr/ingest/api/subscriptions.clj (hunk 6 above)
Comment (eereiter) [reply ingest-app/src/cmr/ingest/api/subscriptions.clj:83]: Oh!! I deleted the wrong comment, and I don't know how to get it back.  Sorry Mitch - he stated: "now this is smart." and I agree with him.

File: common-app-lib/src/cmr/common_app/services/ingest/subscription_common.clj (hunk 7 above)
Comment (mitchstartzel) [reply common-app-lib/src/cmr/common_app/services/ingest/subscription_common.clj:22]: I'm open to discussion on this - my reasoning was that while it _is_ only used in Ingest, it was previously only used in ingest/services while I now needed it in ingest/api. Moving it to common_app/service/ingest seemed like the smart maneuver to cut down on unnecessary imports, since there's no microservice-level common files, but I concede I am not sure if this violates precedent. If it does, I have no issues reverting it.

Comment (ygliuvt) [reply common-app-lib/src/cmr/common_app/services/ingest/subscription_common.clj:22]: Agree with Erich. Functions used only in ingest should stay in ingest code base. `ingest/api` is supposed to call functions within `ingest/services`, so it is logical to put this function in `ingest/services`.

File: ingest-app/src/cmr/ingest/api/subscriptions.clj [hunk 8]
```diff
@@ -60,6 +62,30 @@
            subscriber-id
            concept-id)))))
 
+(defn- search-gran-refs-with-sub-params
+  "Performs a granule search using the provided query parameters. If the query is no good,
+   we throw a service error."
+  [context params]
+  (try
+    (search/find-granule-references context params)
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/api/subscriptions.clj:70]: Since this function is called on every subscription ingest, we should not use the try catch to handle a normal operation scenario. Let's create a new function (maybe called `find-granule-references-return-errors`) to return the error from search regularly rather than throwing exception, and call this new function here.

File: system-int-test/test/cmr/system_int_test/ingest/subscription_ingest_test.clj [hunk 9]
```diff
@@ -167,6 +167,25 @@
...
+                  {:token "mock-echo-system-token"})
+          metadata {:CollectionConceptId (:concept-id coll1)
+                    :Query "options%5Bspatial%5D%5Bor%5D=true"}
+          concept (subscription-util/make-subscription-concept metadata)
+          response (ingest/ingest-concept
+                    concept
+                    {:accept-format :json
+                     :raw? true})
+          {:keys [errors]} (ingest/parse-ingest-body :json response)]
+      (is (re-find #"Error ingesting subscription: invalid query parameters" (first errors))))))
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/ingest/subscription_ingest_test.clj:186]: We should check the error message has the specific error message search-app returned about the invalid query parameter.

File: system-int-test/test/cmr/system_int_test/ingest/subscription_ingest_test.clj [hunk 10]
```diff
@@ -167,6 +167,26 @@
...
+          concept (subscription-util/make-subscription-concept metadata)
+          response (ingest/ingest-concept
+                    concept
+                    {:accept-format :json
+                     :raw? true})
+          {:keys [errors]} (ingest/parse-ingest-body :json response)]
+      (is (re-find #"Error ingesting subscription with query parameters" (first errors)))
+      (is (re-find #"Granule search failed. status: 400 body:" (first errors))))))
+
+
```
Comment (jaybarra) [inline system-int-test/test/cmr/system_int_test/ingest/subscription_ingest_test.clj:189]: extra whitespace nitpick

File: ingest-app/src/cmr/ingest/api/subscriptions.clj [hunk 11]
```diff
@@ -60,6 +62,27 @@
...
 
+(defn- search-gran-refs-with-sub-params
+  "Performs a granule search using the provided query parameters. If the query is no good,
+   we throw a service error."
+  [context params]
+  (if-let [errors (search/find-granule-references-return-errors context params)]
+    (errors/throw-service-error :bad-request
+      (str "Error ingesting subscription with query parameters [" (dissoc params :token) "]: " errors))))
+
+
```
Comment (jaybarra) [inline ingest-app/src/cmr/ingest/api/subscriptions.clj:73]: extra whitespace nitpick

File: ingest-app/src/cmr/ingest/api/subscriptions.clj [hunk 12]
```diff
@@ -60,6 +62,27 @@
            subscriber-id
            concept-id)))))
 
+(defn- search-gran-refs-with-sub-params
+  "Performs a granule search using the provided query parameters. If the query is no good,
+   we throw a service error."
+  [context params]
+  (if-let [errors (search/find-granule-references-return-errors context params)]
```
Comment (jaybarra) [inline ingest-app/src/cmr/ingest/api/subscriptions.clj:69]: There's no 'else' case, so this can be a 'when-let'

**Review by jaybarra — CHANGES_REQUESTED** [review]

File: transmit-lib/src/cmr/transmit/search.clj [hunk 13]
```diff
@@ -89,6 +89,29 @@
...
+        header (ch/context->http-headers context)
+        response (client/post request-url
+                              (merge
+                                (config/conn-params conn)
+                                {:body (codec/form-encode request-body)
+                                 :content-type mt/form-url-encoded
+                                 :throw-exceptions false
+                                 :headers (if token (assoc header config/token-header token) header)}))
+        {:keys [status body]} response]
+    (if (= status 400)
```
Comment (ygliuvt) [inline transmit-lib/src/cmr/transmit/search.clj:111]: this line should be changed to `(when (not= 200 status)`

Comment (ygliuvt) [reply transmit-lib/src/cmr/transmit/search.clj:111]: then you don't need line 113.

File: system-int-test/test/cmr/system_int_test/ingest/subscription_ingest_test.clj [hunk 14]
```diff
@@ -167,6 +167,26 @@
...
+          metadata {:CollectionConceptId (:concept-id coll1)
+                    :Query "options%5Bspatial%5D%5Bor%5D=true"}
+          concept (subscription-util/make-subscription-concept metadata)
+          response (ingest/ingest-concept
+                    concept
+                    {:accept-format :json
+                     :raw? true})
+          {:keys [errors]} (ingest/parse-ingest-body :json response)]
+      (is (re-find #"Error ingesting subscription with query parameters" (first errors)))
+      (is (re-find #"Granule search failed. status: 400 body:" (first errors))))))
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/ingest/subscription_ingest_test.clj:187]: I meant check the exact error message that is returned from search on the validation error. As you can tell already that the whole error message is not very user friendly. The `Granule search failed. status: 400 body:" part is confusing to the end user. That is something only cmr dev cares. To the end user, the message should say something like "Subscription query failed validation with error ....".

File: transmit-lib/src/cmr/transmit/search.clj [hunk 15]
```diff
@@ -89,6 +89,28 @@
...
+        response (client/post request-url
+                              (merge
+                                (config/conn-params conn)
+                                {:body (codec/form-encode request-body)
+                                 :content-type mt/form-url-encoded
+                                 :throw-exceptions false
+                                 :headers (if token (assoc header config/token-header token) header)}))
+        {:keys [status body]} response]
+    (when-not (= status 200)
+      response)))
```
Comment (ygliuvt) [inline transmit-lib/src/cmr/transmit/search.clj:112]: change `response` to `body`.

File: ingest-app/src/cmr/ingest/api/subscriptions.clj [hunk 16]
```diff
@@ -60,6 +62,26 @@
            subscriber-id
            concept-id)))))
 
+(defn- search-gran-refs-with-sub-params
+  "Performs a granule search using the provided query parameters. If the query is no good,
+   we throw a service error."
+  [context params]
+  (when-let [errors (search/find-granule-references-return-errors context params)]
+    (errors/throw-service-error :bad-request
+      (str "Subscription query validation failed with the following error(s): " (:body errors)))))
```
Comment (ygliuvt) [inline ingest-app/src/cmr/ingest/api/subscriptions.clj:71]: change `(:body errors)` to `errors`.

File: transmit-lib/src/cmr/transmit/search.clj [hunk 17]
```diff
@@ -89,6 +89,28 @@
       (errors/internal-error!
         (format "Granule search failed. status: %s body: %s" status body)))))
 
+(defn-timed find-granule-references-return-errors
+  "Find granules by parameters in a post request. If the search fails, returns the failure
```
Comment (ygliuvt) [inline transmit-lib/src/cmr/transmit/search.clj:93]: change to `......, returns the response body.`

File: transmit-lib/src/cmr/transmit/search.clj (hunk 15 above)
Comment (mitchstartzel) [reply transmit-lib/src/cmr/transmit/search.clj:112]: I was thinking that since we don't know what exactly the status code will be (only that it isn't 200) it might be useful to have the whole response, but since I'm not looking for it at all in the validation code, I can't argue with this change.

File: transmit-lib/src/cmr/transmit/search.clj [hunk 18]
```diff
@@ -89,6 +89,27 @@
       (errors/internal-error!
         (format "Granule search failed. status: %s body: %s" status body)))))
 
+(defn-timed find-granule-references-return-errors
```
Comment (ygliuvt) [inline transmit-lib/src/cmr/transmit/search.clj:92]: When I recommended this current name, I wasn't expecting you to change this function in a way that it does not function with valid searches. Since you have modified the function in a way that it does not return any result when the search is valid, we should rename this function to something like `validate-granule-search-params`.

File: transmit-lib/src/cmr/transmit/search.clj [hunk 19]
```diff
@@ -89,6 +89,27 @@
...
+  "Find granules by parameters in a post request. If the search fails, returns the response body."
+  [context params]
+  (let [conn (config/context->app-connection context :search)
+        request-url (str (conn/root-url conn) "/granules.json")
```
Comment (ygliuvt) [inline transmit-lib/src/cmr/transmit/search.clj:96]: change `/granules.json` to `/granules`.

File: transmit-lib/src/cmr/transmit/search.clj [hunk 20]
```diff
@@ -89,6 +89,27 @@
...
+        request-body (-> params
+                         (dissoc :token)
+                         (assoc :page_size 0))
```
Comment (ygliuvt) [inline transmit-lib/src/cmr/transmit/search.clj:99]: Don't expect you put this here, but if we change the function name, we are ok.

File: transmit-lib/src/cmr/transmit/search.clj [hunk 21]
```diff
@@ -89,6 +89,27 @@
       (errors/internal-error!
         (format "Granule search failed. status: %s body: %s" status body)))))
 
+(defn-timed find-granule-references-return-errors
+  "Find granules by parameters in a post request. If the search fails, returns the response body."
```
Comment (ygliuvt) [inline transmit-lib/src/cmr/transmit/search.clj:93]: update the docstring to reflect what this function is doing now.

**Review by ygliuvt — APPROVED** [review]

**Review by jaybarra — APPROVED** [review]
